pkg/util/indexedlist/indexedlist.go

```diff
@@ -0,0 +1,210 @@
+package indexedlist
+
+import (
+	"fmt"
+	"sync"
+
+	tt "github.com/filecoin-project/mir/pkg/trantor/types"
+)
+
+type element[K comparable, V any] struct {
+	prev *element[K, V]
+	next *element[K, V]
+	key  K
+	val  V
+}
+
+// IndexedList is a doubly-linked list with a map index.
+// Appending, prepending, and removal of arbitrary elements is in constant time and thread-safe.
```

Comment:
Appending, preprending and removing ~of~ ?

Reply:
It was `removal of arbitrary elements`, but I changed that sentence now to sound a bit better.